### ISSUE TITLE
ignoreDifferences does not work with helm generated Secrets

### ISSUE THREAD

**Issue by adyanth** [argoproj/argo-cd#8810]
Checklist:

* [X] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [X] I've included steps to reproduce the bug.
* [X] I've pasted the output of `argocd version`.

**Describe the bug**

Even with the `ignoreDifferences`, argocd does not ignore differences.
Usecase: Gitea helm chart has a random generator for the PostgreSQL password. This only needs to happen once. ArgoCD considers this as a diff and updates the secret every time (losing the original password in the process too).

**To Reproduce**

Apply the below yaml using app of apps, generate a dummy commit and refresh.

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: gitea
  namespace: argocd
spec:
  destination:
    namespace: gitea
    server: "https://kubernetes.default.svc"
  source:
    repoURL: "https://gitea.com/gitea/helm-chart.git"
    path: .
    targetRevision: HEAD
  project: default
  syncPolicy:
    syncOptions:
      - CreateNamespace=true
    automated:
      prune: false
      selfHeal: false
  ignoreDifferences:
    - group: core
      kind: Secret
      name: gitea-postgresql
      namespace: gitea # Tried with and without, is this really needed?
      jsonPointers:
        - /data

```

**Expected behaviour**

I expect the secret to not be updated.

**Screenshots**
<img width="928" alt="Screenshot 2022-03-17 at 5 54 49 PM" src="https://user-images.githubusercontent.com/33192449/158808491-c0fd798f-7a85-4786-98f3-1efa934808eb.png">

<img width="938" alt="Screenshot 2022-03-17 at 5 55 05 PM" src="https://user-images.githubusercontent.com/33192449/158808503-a2bf2751-41a7-4256-9aeb-55b1abeb1886.png">

**Version**

```shell
argocd: v2.2.5+8f981cc
  BuildDate: 2022-02-05T01:33:25Z
  GitCommit: 8f981ccfcf942a9eb00bc466649f8499ba0455f5
  GitTreeState: clean
  GoVersion: go1.16.11
  Compiler: gc
  Platform: linux/amd64
```

**Logs**

```
N/A
```

**Comment by wizardofdevops** [comment]
having same issue --- did you find a solution or workaround?

**Comment by adyanth** [comment]
Nothing yet, I just live with it for now :*)

**Comment by bmarick** [comment]
any update on this issue?

**Comment by tsub** [comment]
I removed `group: core` and it works in my environment. :smile:

```diff
--- before	2022-10-28 11:24:17.000000000 +0900
+++ after	2022-10-28 11:24:25.000000000 +0900
@@ -19,8 +19,7 @@
       prune: false
       selfHeal: false
   ignoreDifferences:
-    - group: core
-      kind: Secret
+    - kind: Secret
       name: gitea-postgresql
       namespace: gitea # Tried with and without, is this really needed?
       jsonPointers:
```

Secret's API group should be `core`, but it may not be necessary to specify `core`. 🤔

**Comment by mindlace** [comment]
**TL;DR** - don't specify data: keys in your template if you want to change them at runtime

This is happening because argocd is doing a structural diff not a values one.

Say you have a template secret with dummy values for the items you want to replace, like so:
```yaml
apiVersion: v1
data:
  client_id: dGVtcAo=
  client_secret: dGVtcAo=
kind: Secret
metadata:
  name: my-app-secret
type: Opaque
```
From argocd's perspective, the contents of `data:` is a mapping that contains two keys: `client_id` and `client_secret`. 

If you overwrite the *values* of `client_id` or `client_secret`, you haven't changed the data mapping: it's still a mapping with two keys.

Therefore you need to avoid specifying keys that you want to overwrite, and it will work.

Alternately you can specify your sub-keys in `ignoreDifferences`:
```yaml
jsonPointers:
  - /data/client_id
  - /data/client_secret
```

**Comment by mindlace** [comment]
Adding [RespectIgnoreDifferences=true](https://argo-cd.readthedocs.io/en/stable/user-guide/sync-options/#respect-ignore-difference-configs) solved the problem on our end.

**Comment by lucasmaurice** [comment]
Same issue for me. It looks like the number of "+" is different on each sides so their is maybe an issue with the redacted secrets.

![CleanShot 2024-08-17 at 09 48 21@2x](https://github.com/user-attachments/assets/8876489b-1fbf-4d1f-b93d-0c2eb1ec5593)

**Comment by urcus** [comment]
Hello, having same issue i've tryed the different solution proposed here
As my syncPolicy is set as:

```yaml
syncPolicy:
   syncOptions:
     - RespectIgnoreDifferences=true
```
    
 and my ignoreDifferences as :
 
```yaml
ignoreDifferences:
  - kind: Secret
    name: app-ca
    namespace: kube-system
    jsonPointers:
       - /data
 ```
 But i still have my secret regenerated at every sync.

**Comment by rmn-lux** [comment]
+1 the same problem ^

ArgoCD v2.13.3

**Comment by JavierPorron** [comment]
Same problem 
ArgoCD v2.14.3

**Comment by bitnik** [comment]
In our chart the secret was defined with `stringData`. Our Application was configured to ignore only `/data` and this was the problem. Adding `/stringData` into the list solved the issuer for us.

**Comment by shsamkit** [comment]
```
      syncPolicy:
        ...
        syncOptions:
        ...
          - RespectIgnoreDifferences=true # IMPORTANT
      ignoreDifferences:
        - kind: Secret
          name: $RESOURCE_NAME
          namespace: $NAMESPACE # ignorable
          jsonPointers: 
            - /data # IMPORTANT
```

This enabled ignoreDifferences for my setup.  [ref. for ignoreDifferences](https://argo-cd.readthedocs.io/en/stable/user-guide/diffing/#application-level-configuration), [ref. for RespectIgnoreDifferences](https://argo-cd.readthedocs.io/en/latest/user-guide/sync-options/#respect-ignore-differences-configs)
